server/driver/weights_list.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * WiVRn VR streaming
+ * Copyright (C) 2024  galister <galister@livrevr.org>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "clock_offset.h"
+#include "history.h"
+
+#include <cmath>
+
+template <size_t NumWeights>
+class weights_list : public history<weights_list<NumWeights>, std::array<float, NumWeights>>
+{
+public:
+	static std::array<float, NumWeights> interpolate(const std::array<float, NumWeights> & a, const std::array<float, NumWeights> & b, float t)
+	{
+		std::array<float, NumWeights> result;
+		for (size_t i = 0; i < NumWeights; i++)
+		{
+			result[i] = std::clamp(std::lerp(a[i], b[i], t), 0.0f, 1.0f);
+		}
+		return result;
+	}
+
+	static std::array<float, NumWeights> extrapolate(const std::array<float, NumWeights> & a, const std::array<float, NumWeights> & b, uint64_t ta, uint64_t tb, uint64_t t)
+	{
+		float h = (tb - ta) / 1.e9;
+
+		std::array<float, NumWeights> result = t < ta ? a : b;
+		for (size_t i = 0; i < NumWeights; i++)
+		{
+			auto vel = (b[i] - a[i]) / h;
+			float dt = (t - tb) / 1.e9;
+
+			result[i] = std::clamp(result[i] + vel * dt, 0.0f, 1.0f);
+		}
+		return result;
+	}
```

Comment:
It should be possible to not implement this by replacing  the `if (extrapolate)` with `if constexpr (extrapolate)`.

I don't know how much sense it makes to extrapolate expressions.

Reply:
it's not being used right now as the underlying history has extrapolate off, but I implemented it in case it would be used for something else that needs extrapolation